### PR TITLE
dashboard: ntt total supply on evm vs total locked on sol

### PR DESCRIPTION
**Pull request by bingyuyap**
This PR does a few things:
- moves `ntt_rate_limit` and `ntt_total_supply_locked` rpc invocation from frontend to cloud function due to Solana Mainnet RPC CORS gating
  - `computeNTT<something>`: invokes RPC and saves data into a bucket cache
  - `getNTT<something>`: takes data from bucket cache and return to client
- getNTTRateLimits returns 2 types of format to minimize code logic duplication:
  - Prometheus text format by default
  - JSON - if request `Content-Type` is `application/json`
- Note that `getNTTRateLimits` is gated by a fake bearer token because Grafana Metrics Endpoint Integration mandates the URL to be gated
- adds a table to show number of tokens locked on Solana vs total number of tokens on EVM chains

image

### CODE REVIEW

File: common/src/solana.ts [hunk 1]
```diff
@@ -50,3 +53,64 @@ export const getSolanaTokenDecimals = async (rpc: string, mintAddress: string):
...
+  if (response.data.error) {
+    throw new Error(`Error fetching ${method} account: ${response.data.error.message}`);
+  }
+  return response.data.result;
+}
+
+// Helper function to decode base64 data
+export function decodeBase64Data(encodedData: string) {
+  return new Uint8Array(
+    atob(encodedData)
```
Comment (panoel) [inline common/src/solana.ts:112]: This function is deprecated.  Can we not use Buffer.from(encodedData, "base64"), or something like that, instead?

File: cloud_functions/src/computeNTTRateLimits.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+        const inboundCapacity = await Promise.all(inboundCapacityPromises);
+
+        return {
+          tokenName: token,
+          inboundCapacity: inboundCapacity,
+        };
+      })
+    );
+    await Promise.all(rateLimitPromises);
```
Comment (panoel) [inline cloud_functions/src/computeNTTRateLimits.ts:126]: Do you still have to do this await even though there is an await on line 110?

File: cloud_functions/src/computeNTTRateLimits.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+let bucketName: string = 'wormhole-ntt-cache';
+if (assertEnvironmentVariable('NETWORK') === 'Testnet') {
+  bucketName = 'wormhole-ntt-cache-testnet';
+}
+
+const cacheBucket = storage.bucket(bucketName);
+const cacheFileName = 'ntt-rate-limits-cache.json';
+const cloudStorageCache = cacheBucket.file(cacheFileName);
+
+async function computeNTTRateLimits_(network: Network, token: string, chain: Chain) {
```
Comment (panoel) [inline cloud_functions/src/computeNTTRateLimits.ts:28]: Typescript question: does this function need to return Promise<void> ?

**Review by panoel — COMMENTED** [review]
I did not review the components files as I'm not sure how they work.

File: common/src/solana.ts (hunk 1 above)
Comment (bingyuyap) [reply common/src/solana.ts:112]: Updated

File: cloud_functions/src/computeNTTRateLimits.ts (hunk 2 above)
Comment (bingyuyap) [reply cloud_functions/src/computeNTTRateLimits.ts:126]: You are right! Changed this

File: cloud_functions/src/computeNTTRateLimits.ts (hunk 3 above)
Comment (bingyuyap) [reply cloud_functions/src/computeNTTRateLimits.ts:28]: by default async functions return Promises, but I will add a Promise<NTTRateLimit> for enforcing type checking

**Review by panoel — APPROVED** [review]

**Pull request merged**